File: scilab/routines/system/expr.f
      subroutine expr
c     
c     analyseur d'expressions
c     
c     Copyright INRIA
      INCLUDE '../stack.h'
      parameter (nz1=nsiz-1,nz2=nsiz-2)
      integer iadr,sadr
c     
      integer op,r,blank,sign,plus,minus,name,colon,eye(nsiz),ou,et
      integer equal,less,great,not,eol,p,temp
      logical eptover, skip, compil, ifexpr, istrue
      
      external gettype,istrue
      
      data colon/44/,blank/40/,plus/45/,minus/46/,ou/57/,et/58/
      data equal/50/,less/59/,great/60/,not/61/
      data eye/672014862,nz1*673720360/,name/1/,eol/99/
c     

      sadr(l)=(l/2)+1
      iadr(l)=l+l-1
      if (ddt .eq. 4) then
         write(buf(1:8),'(2i4)') pt,rstk(pt)
         call basout(io,wte,' expr   pt:'//buf(1:4)//' rstk(pt):'
     &        //buf(5:8))
      endif
c     
      r = rstk(pt)
      if(r.eq.204) goto 85
      ir=r/100
      if(ir.ne.1) goto 01
      goto(05,06,25,26,61,73,74,82,83,86,87,102,104,105,22),r-100
c     
 01   if(sym.ge.ou.and.sym.le.great) then
         call error(40)
         return
      endif
 02   continue
      kount = 1
      if(sym.eq.not) goto 70
      if (sym .eq. colon) call putid(syn(1),eye)
c      if (sym .eq. colon) sym = name
 03   sign = plus
      if (sym .eq. minus) sign = minus
      if (sym.eq.plus .or. sym.eq.minus) call getsym
      if ( eptover(1,psiz-1)) return
      pstk(pt) = sign + 256*kount
      rstk(pt) = 101
      icall=2
c     *call* term
      return
 05   sign = mod(pstk(pt),256)
      kount = pstk(pt)/256
      pt = pt-1
      if (sign .ne. minus) goto 10
      rhs=1
      pt=pt+1
      pstk(pt)=kount
      fin=minus
      rstk(pt)=102
      icall=4
c     *call* allops(minus)
      return
 06   kount=pstk(pt)
      pt=pt-1
 10   if (sym.eq.plus .or. sym.eq.minus) go to 20
      if(rstk(pt).eq.113) goto 104
      if(rstk(pt).eq.112) goto 102
      if(sym.eq.eol) goto 50
      if(sym.ge.ou.or.sym.eq.equal) goto 70
      go to 50
 20   if (rstk(pt) .ne. 301) go to 21
c     blank or tab is delimiter inside angle brackets
      ls = lpt(3) - 2
      if (abs(lin(ls)).eq.blank.and.abs(lin(lpt(3))).ne.blank) go to 50
 21   op = sym
      call getsym
      pt=pt+1
      pstk(pt) = op + 256*kount
      if(sym.ne.not) goto 23
      rstk(pt)=115
c     *call* lfact
      goto 85
 22   goto 25
c
 23   rstk(pt) = 103
      icall=2
c     *call* term
      return
 25   op = mod(pstk(pt),256)
      kount = pstk(pt)/256
      rhs=2
      pstk(pt)=kount
      rstk(pt)=104
      fin=op
      icall=4
c     *call* allops(op)
      return
 26   kount=pstk(pt)
      pt=pt-1
      go to 10
 50   if (sym .ne. colon) go to 60
      call getsym
      kount = kount+1
      go to 03
 60   if (kount .gt. 3) then
         call error(33)
         if (err .gt. 0) return
      endif
      rhs = kount
      if (kount .le. 1) return
      pt=pt+1
      rstk(pt)=105
      fin=colon
      icall=4
c     *call* allops(colon)
      return
 61   pt=pt-1
      r = rstk(pt)
      ir=r/100
      if(ir.ne.1) return
      goto(05,06,25,26,61,73,74,82,83,86,87,102,104,105),r-100
c     
c     in-line lexpr
 70   continue
      if (ddt .eq. 4) then
         write(buf(1:8),'(2i4)') pt,rstk(pt)
         call basout(io,wte,' lexpr  pt:'//buf(1:4)//' rstk(pt):'
     &        //buf(5:8))
      endif
 72   continue
      if (eptover(1,psiz-1)) return
      ids(1,pt)=0
      ids(2,pt)=err1
      if(sym.eq.ou) then
c     call getsym
         pstk(pt)=ou
         if(ifexpr()) then
c     .     checking for possible logical 'if expression' 
c     .     evaluation shortcircuit
            if(comp(1).ne.0) then
               if (compil(30,1,0,0,0)) then 
                  if (err.gt.0) return
                  ids(1,pt)=comp(1)
               endif
            else
               temp=abs(istk(iadr(lstk(top))))
               if (temp .ne.8 ) then
                  if ( istrue(0)) then
c     .              first term is true there is no use to evaluate the other
                     ids(1,pt)=1
c     .              err1 <>0 sets interpretation without evaluation
                     err1=1
                  endif
               endif
            endif
         endif
      else
         pstk(pt)=0
      endif
      pstk(pt)=pstk(pt)+256*kount
      rstk(pt)=106
c     *call* lterm
      goto 80
 73   continue
      op = mod(pstk(pt),256)
      kount = pstk(pt)/256
      pt=pt-1
      if(op.eq.0) goto 75
      if(comp(1).eq.0.and.ids(1,pt+1).eq.1) then
c     . term has not been evaluated
         err1=ids(2,pt+1)
         goto 75
      endif
      icall=4
      fin=ou
      rhs=2
      pt=pt+1
      rstk(pt)=107
      pstk(pt)=kount
c     *call* allops(ou)
      return
 74   continue
      kount=pstk(pt)
      if(comp(1).ne.0.and.ids(1,pt).ne.0) then
         if (compil(30,0,ids(1,pt)-1,0,0)) then 
            if (err.gt.0) return
         endif
      endif
      pt=pt-1
 75   if(sym.eq.ou) goto 72
      goto 50
c     
c     in-line lterm
 80   continue
      if (ddt .eq. 4) then
         write(buf(1:8),'(2i4)') pt,rstk(pt)
         call basout(io,wte,' lterm   pt:'//buf(1:4)//' rstk(pt):'
     &        //buf(5:8))
      endif
 81   continue
      if (eptover(1,psiz-1))  return
      ids(1,pt)=0
      ids(2,pt)=err1
      if(sym.eq.et) then
         pstk(pt)=et
         ids(1,pt)=0
         if(ifexpr()) then
c          . if expression evaluation, checking for possible
c          . logical expression evaluation shortcircuit
            if(comp(1).ne.0) then
               if (compil(30,0,0,0,0)) then 
                  if (err.gt.0) return
                  ids(1,pt)=comp(1)
               endif
            else
                temp=abs(istk(iadr(lstk(top))))
                if (temp .ne.8) then
                  if (.not.istrue(0)) then
c     .             first term is false there is no use to evaluate the other
                    ids(1,pt)=1
c     .             err1 <>0 sets interpretation without evaluation
                    err1=1
                  endif
                endif
            endif
         endif
      else
         pstk(pt)=0
      endif
      pstk(pt)=pstk(pt)+256*kount
      rstk(pt)=108
c     *call* lfact
      goto 85
 82   continue
      op = mod(pstk(pt),256)
      kount = pstk(pt)/256
      pt=pt-1
      if(op.eq.0) goto 84
      if(comp(1).eq.0.and.ids(1,pt+1).eq.1) then
c     . term has not been evaluated
         err1=ids(2,pt+1)
         goto 84
      endif

      icall=4
      fin=et
      rhs=2
      pt=pt+1
      pstk(pt)=kount
      rstk(pt)=109
c     *call* allops(et)
      return
 83   continue
      kount=pstk(pt)
      if(comp(1).ne.0.and.ids(1,pt).ne.0) then
         if (compil(30,0,ids(1,pt)-1,0,0)) then 
            if (err.gt.0) return
         endif
      endif
      pt=pt-1
 84   if(sym.ne.et) goto 73
      goto 81
c     
c     in-line lfact
c     
 85   continue
      if (ddt .eq. 4) then
         write(buf(1:8),'(2i4)') pt,rstk(pt)
         call basout(io,wte,' lfact  pt:'//buf(1:4)//' rstk(pt):'
     &        //buf(5:8))
      endif
      if ( eptover(1,psiz-1)) return
      pstk(pt)=0
      if(sym.eq.not.and.char1.ne.equal) then
         pstk(pt)=not
         call getsym
      endif
      pstk(pt)=pstk(pt)+256*kount	  
      rstk(pt)=110
c     *call* lprim
      goto 100
 86   continue
      op = mod(pstk(pt),256)
      kount = pstk(pt)/256
      pt=pt-1
      if(op.eq.0) goto 82
      fin=op
      rhs=1
      pt=pt+1
      pstk(pt)=kount
      rstk(pt)=111
      icall=4
c     *call* allops(not)
      return
 87   kount=pstk(pt)
      pt=pt-1
C     next two lines to handle a+~b and a*~b,...
      if(rstk(pt).eq.115) goto 22
      if(rstk(pt).eq.204) return
      goto 82
c     
c     in-line lprim
c     
 100  continue
      if (ddt .eq. 4) then
         write(buf(1:8),'(2i4)') pt,rstk(pt)
         call basout(io,wte,' lprim  pt:'//buf(1:4)//' rstk(pt):'
     &        //buf(5:8))
      endif
      if(mod(pstk(pt),256).ne.0) goto 101
      if(sym.ne.et.and.sym.ne.ou) goto 103
      call getsym
C modif SS
C      if(sym.ge.ou.or.sym.eq.equal) goto 70
      if ( eptover(0,psiz-1)) return
 101  pt=pt+1
      rstk(pt)=112
      icall=1
c     *call* expr
      goto 02
 102  pt=pt-1
      if(sym.ne.equal.and.sym.lt.less.or.sym.eq.eol)  goto 86
 103  op=sym
      call getsym
c      if(rstk(pt-3).ne.803) then
         if(op.eq.equal.and.sym.ne.equal) then
            call msgs(7,0)
         endif
c      endif

      if(sym.eq.equal.or.sym.eq.great) then
         if(op.ne.equal) op=op+sym
         call  getsym
         if(op.eq.not+equal) op=less+great
      endif
      pt=pt+1
      rstk(pt)=113
      pstk(pt)=op
      icall=1
c     *call* expr
      goto 02
 104  fin=pstk(pt)
      rhs=2
      rstk(pt)=114
      icall=4
c     *call* allops(fin)
      return
 105  continue
      goto 102
c     
c 999  call error(22)
c      if (err .gt. 0) return
c      return
      end

      logical function ifexpr()
      INCLUDE '../stack.h'
      integer p
      integer blank
      data blank/40/
      if(err1.ne.0) goto 20
      p=pt
 10   p=p-1
c     the test rstk(p).eq.307.and.ids(1,p-1).ne.blank) checks if we are
c     evaluating the argument of a function. ids(1,p-1).ne.blank
C     excludes the case of a parenthesized if expression
      if (p.eq.0.or.(rstk(p).eq.307.and.ids(1,p-1).ne.blank)) goto 20
      if (rstk(p).ne.803) goto 10
      ifexpr=.true.
      return
 20   ifexpr=.false.
      return
      end

File: scilab/modules/core/src/c/expr.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2007 - INRIA - Allan CORNET
 * 
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at    
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 *
 */
#include "expr.h"
#include "ifexpr.h"
#include "msgs.h"
#include "stack-def.h"
#include "stack-c.h"
#include "Scierror.h"
/*--------------------------------------------------------------------------*/ 
static int inc = 1;
static int checkvalue = 4095;
static int code = 30;
static int val = 0;
/*--------------------------------------------------------------------------*/ 
extern int C2F(compil)(int *, int *, int *, int *, int *);
extern int C2F(getsym)(void);
extern int C2F(istrue)(int *);
extern int C2F(eptover)(int *, int *);
/*--------------------------------------------------------------------------*/ 
#define  et      58 /* @TODO does 'et' is 'and' in english ? */
#define  equal   50
#define  less    59
#define  great   60
#define  not     61
#define  eol     99
#define  colon   44
#define  blank   40
#define  plus    45
#define  minus   46
#define  ou      57 /* @TODO does 'ou' is 'or' in english ? */
/*--------------------------------------------------------------------------*/ 
int C2F(expr)(void)
{
	static int eye[6] = { 672014862,673720360,673720360,673720360, 673720360,673720360 };
	int i = 0, j = 0;
	int r = 0 , s = 0, ir = 0, op = 0, ls = 0, sign = 0;
	int temp = 0;
	int kount = 0;

	if (C2F(iop).ddt == 4) { }

	r = C2F(recu).rstk[C2F(recu).pt - 1];
	if (r == 204) goto L85;

	ir = r / 100;

	if (ir != 1) goto L1;
	
	switch (r - 100) 
	{
		case 1:  goto L5;
		case 2:  goto L6;
		case 3:  goto L25;
		case 4:  goto L26;
		case 5:  goto L61;
		case 6:  goto L73;
		case 7:  goto L74;
		case 8:  goto L82;
		case 9:  goto L83;
		case 10:  goto L86;
		case 11:  goto L87;
		case 12:  goto L102;
		case 13:  goto L104;
		case 14:  goto L102;
		case 15:  goto L23;
	}

L1:
	if (C2F(com).sym >= ou && C2F(com).sym <= great) 
	{
		int code_error = 40;
		Error(code_error);
		return 0;
	}
L2:
	kount = 1;
	if (C2F(com).sym == not) goto L70;
	
	if (C2F(com).sym == colon) C2F(putid)(C2F(com).syn, eye);

L3:
	s = 1;
L4:
	if (C2F(com).sym == minus) s = -s;
	
	if (C2F(com).sym == plus || C2F(com).sym == minus) 
	{
		C2F(getsym)();
		goto L4;
	}
	sign = plus;
	if (s < 0) sign = minus;
	
	if (C2F(eptover)(&inc, &checkvalue)) return 0;
	
	C2F(recu).pstk[C2F(recu).pt - 1] = sign + (kount << 8);
	C2F(recu).rstk[C2F(recu).pt - 1] = 101;
	C2F(recu).icall = 2;
	/* *call* term */
	return 0;
L5:
	sign = C2F(recu).pstk[C2F(recu).pt - 1] % 256;
	kount = C2F(recu).pstk[C2F(recu).pt - 1] / 256;
	--C2F(recu).pt;
	if (sign != minus) goto L10;
	
	Rhs = 1;
	++C2F(recu).pt;
	C2F(recu).pstk[C2F(recu).pt - 1] = kount;
	Fin = minus;
	C2F(recu).rstk[C2F(recu).pt - 1] = 102;
	C2F(recu).icall = 4;
	/* *call* allops(minus) */
	return 0;
L6:
	kount = C2F(recu).pstk[C2F(recu).pt - 1];
	--C2F(recu).pt;
L10:
	if (C2F(com).sym == plus || C2F(com).sym == minus) goto L20;
	if (C2F(recu).rstk[C2F(recu).pt - 1] == 113) goto L104;
	if (C2F(recu).rstk[C2F(recu).pt - 1] == 112) goto L102;
	if (C2F(com).sym == eol) goto L50;
	if (C2F(com).sym >= ou || C2F(com).sym == equal) goto L70;
	goto L50;
L20:
	if (C2F(recu).rstk[C2F(recu).pt - 1] != 301) goto L21;
	/* blank or tab is delimiter inside angle brackets */
	ls = C2F(iop).lpt[2] - 2;
	if ( (i = C2F(iop).lin[ls - 1], abs(i)) == blank && (j = C2F(iop).lin[C2F(iop).lpt[2] - 1], abs(j)) != blank) goto L50;
		
L21:
	op = C2F(com).sym;
	C2F(getsym)();
	/* next lines added to handle sequence of + and - operators, S. Steer */
	/* 03/2005 (Matlab compatibility). Here i implemented a lazy way */
	/* without calling unary + or unary - operator */
L22:
	if (C2F(com).sym == plus) 
	{
		/* 1++2 or 1-+2 */
		C2F(getsym)();
		goto L22;
	}
	if (C2F(com).sym == minus) 
	{
		/* 1+-2 or 1--2 */
		if (op == minus) op = plus;
		else op = minus;
			
		C2F(getsym)();
		goto L22;
	}
	++C2F(recu).pt;
	C2F(recu).pstk[C2F(recu).pt - 1] = op + (kount << 8);
	if (C2F(com).sym != not) goto L24;
		
	C2F(recu).rstk[C2F(recu).pt - 1] = 115;
	/* *call* lfact */
	goto L85;

L23:
	goto L25;

L24:
	C2F(recu).rstk[C2F(recu).pt - 1] = 103;
	C2F(recu).icall = 2;
	/* *call* term */
	return 0;
L25:
	op = C2F(recu).pstk[C2F(recu).pt - 1] % 256;
	kount = C2F(recu).pstk[C2F(recu).pt - 1] / 256;
	Rhs = 2;
	C2F(recu).pstk[C2F(recu).pt - 1] = kount;
	C2F(recu).rstk[C2F(recu).pt - 1] = 104;
	Fin = op;
	C2F(recu).icall = 4;
	/* *call* allops(op) */
	return 0;
L26:
	kount = C2F(recu).pstk[C2F(recu).pt - 1];
	--C2F(recu).pt;
	goto L10;
L50:
	if (C2F(com).sym != colon) goto L60;

	C2F(getsym)();
	++kount;
	goto L3;
L60:
	if (kount > 3) 
	{
		int code_error = 33;
		Error(code_error);
		if (Err > 0) return 0;
	}
	Rhs = kount;
	if (kount <= 1) return 0;
	++C2F(recu).pt;
	C2F(recu).rstk[C2F(recu).pt - 1] = 105;
	Fin = colon;
	C2F(recu).icall = 4;
	/* *call* allops(colon) */
	return 0;
L61:
	--C2F(recu).pt;
	r = C2F(recu).rstk[C2F(recu).pt - 1];
	ir = r / 100;
	if (ir != 1) return 0;

	switch (r - 100) 
	{
		case 1:  goto L5;
		case 2:  goto L6;
		case 3:  goto L25;
		case 4:  goto L26;
		case 5:  goto L61;
		case 6:  goto L73;
		case 7:  goto L74;
		case 8:  goto L82;
		case 9:  goto L83;
		case 10:  goto L86;
		case 11:  goto L87;
		case 12:  goto L102;
		case 13:  goto L104;
		case 14:  goto L102;
	}

	/* in-line lexpr */
L70:
	if (C2F(iop).ddt == 4) {}
L72:
	if (C2F(eptover)(&inc, &checkvalue)) return 0;
	C2F(recu).ids[C2F(recu).pt * 6 - 6] = 0;
	C2F(recu).ids[C2F(recu).pt * 6 - 5] = C2F(errgst).err1;
	if (C2F(com).sym == ou) 
	{
		/* call getsym */
		C2F(recu).pstk[C2F(recu).pt - 1] = ou;
		if (C2F(ifexpr)()) 
		{
			/* checking for possible logical 'if expression' */
			/* evaluation shortcircuit */
			if (C2F(com).comp[0] != 0) 
			{
				if (C2F(compil)(&code, &inc, &val, &val, &val)) 
				{
					if (Err > 0) return 0;
					C2F(recu).ids[C2F(recu).pt * 6 - 6] = C2F(com).comp[0];
				}
			} 
			else 
			{
				temp = (i = *istk(iadr(C2F(vstk).lstk[Top - 1])), abs(i));
				if ( (temp != 8) && (C2F(istrue)(&val)) )
				{
					/* first term is true there is no use to evaluate the other */
					C2F(recu).ids[C2F(recu).pt * 6 - 6] = 1;
					/* err1 <>0 sets interpretation without evaluation */
					/* use special value to be able to distinguish from */
					/* recovered errors */
					C2F(errgst).err1 = 9191919;
				}
			}
		}
	}
	else C2F(recu).pstk[C2F(recu).pt - 1] = 0;
	C2F(recu).pstk[C2F(recu).pt - 1] += kount << 8;
	C2F(recu).rstk[C2F(recu).pt - 1] = 106;
	/* *call* lterm */
	goto L80;
L73:
	op = C2F(recu).pstk[C2F(recu).pt - 1] % 256;
	kount = C2F(recu).pstk[C2F(recu).pt - 1] / 256;
	--C2F(recu).pt;
	if (op == 0) goto L75;
	
	if (C2F(com).comp[0] == 0 && C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 6] == 1) 
	{
		/* term has not been evaluated */
		if ((i = -C2F(errgst).errct, abs(i)) / 100000 == 0) 
		{
			C2F(errgst).err1 = C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 5];
		} 
		else if (C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 5] != 0) 
		{
			C2F(errgst).err1 = C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 5];
		} 
		else 
		{
			if (C2F(errgst).err1 == 9191919) C2F(errgst).err1 = 0;
		}
		if (C2F(errgst).err1 > 0) return 0;
		goto L75;
	}
	C2F(recu).icall = 4;
	Fin = ou;
	Rhs = 2;
	++C2F(recu).pt;
	C2F(recu).rstk[C2F(recu).pt - 1] = 107;
	C2F(recu).pstk[C2F(recu).pt - 1] = kount;
	/* *call* allops(ou) */
	return 0;
L74:
	kount = C2F(recu).pstk[C2F(recu).pt - 1];
	if (C2F(com).comp[0] != 0 && C2F(recu).ids[C2F(recu).pt * 6 - 6] != 0) 
	{
		i = C2F(recu).ids[C2F(recu).pt * 6 - 6] - 1;
		if ( (C2F(compil)(&code, &val, &i, &val, &val)) && (Err > 0) ) return 0;
	}
	--C2F(recu).pt;
L75:
	if (C2F(com).sym == ou) goto L72;
	goto L50;
	/* in-line lterm */
L80:
	if (C2F(iop).ddt == 4) { }
L81:
	if (C2F(eptover)(&inc, &checkvalue)) return 0;
	
	C2F(recu).ids[C2F(recu).pt * 6 - 6] = 0;
	C2F(recu).ids[C2F(recu).pt * 6 - 5] = C2F(errgst).err1;
	if (C2F(com).sym == et) 
	{
		C2F(recu).pstk[C2F(recu).pt - 1] = et;
		C2F(recu).ids[C2F(recu).pt * 6 - 6] = 0;
		if (C2F(ifexpr)()) 
		{
			/* if expression evaluation, checking for possible */
			/* logical expression evaluation shortcircuit */
			if (C2F(com).comp[0] != 0) 
			{
				if (C2F(compil)(&code, &val, &val, &val, &val)) 
				{
					if (Err > 0) return 0;
					C2F(recu).ids[C2F(recu).pt * 6 - 6] = C2F(com).comp[0];
				}
			} 
			else 
			{
				temp = (i = *istk(iadr(C2F(vstk).lstk[Top - 1])), abs(i));
				if ( (temp != 8) && (! C2F(istrue)(&val)) )
				{
					/* first term is false there is no use to evaluate the other */
					C2F(recu).ids[C2F(recu).pt * 6 - 6] = 1;
					/* err1 <>0 sets interpretation without evaluation */
					C2F(errgst).err1 = 9191919;
				}
			}
		}
	} 
	else 
	{
		C2F(recu).pstk[C2F(recu).pt - 1] = 0;
	}
	C2F(recu).pstk[C2F(recu).pt - 1] += kount << 8;
	C2F(recu).rstk[C2F(recu).pt - 1] = 108;
	/* *call* lfact */
	goto L85;
L82:
	op = C2F(recu).pstk[C2F(recu).pt - 1] % 256;
	kount = C2F(recu).pstk[C2F(recu).pt - 1] / 256;
	--C2F(recu).pt;
	if (op == 0) goto L84;
	
	if (C2F(com).comp[0] == 0 && C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 6] == 1) 
	{
		/* term has not been evaluated */
		if ((i = -C2F(errgst).errct, abs(i)) / 100000 == 0) 
		{
			C2F(errgst).err1 = C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 5];
		} 
		else if (C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 5] != 0) 
		{
			/* error detected before if expression evaluation (should not occur ?) */
			C2F(errgst).err1 = C2F(recu).ids[(C2F(recu).pt + 1) * 6 - 5];
		} 
		else 
		{
			if (C2F(errgst).err1 == 9191919) C2F(errgst).err1 = 0;
		}
		if (C2F(errgst).err1 > 0) return 0;
		goto L84;
	}
	C2F(recu).icall = 4;
	Fin = et;
	Rhs = 2;
	++C2F(recu).pt;
	C2F(recu).pstk[C2F(recu).pt - 1] = kount;
	C2F(recu).rstk[C2F(recu).pt - 1] = 109;
	/* *call* allops(et) */
	return 0;
L83:
	kount = C2F(recu).pstk[C2F(recu).pt - 1];
	if (C2F(com).comp[0] != 0 && C2F(recu).ids[C2F(recu).pt * 6 - 6] != 0) 
	{
		i = C2F(recu).ids[C2F(recu).pt * 6 - 6] - 1;
		if ( (C2F(compil)(&code, &val, &i, &val, &val)) && (Err > 0) ) return 0;
	}
	--C2F(recu).pt;
L84:
	if (C2F(com).sym != et) goto L73;
	goto L81;

	/* in-line lfact */
L85:
	if (C2F(iop).ddt == 4) { }
	if (C2F(eptover)(&inc, &checkvalue)) return 0;
		
	C2F(recu).pstk[C2F(recu).pt - 1] = 0;
	if (C2F(com).sym == not && C2F(com).char1 != equal) 
	{
		C2F(recu).pstk[C2F(recu).pt - 1] = not;
		C2F(getsym)();
	}
	C2F(recu).pstk[C2F(recu).pt - 1] += kount << 8;
	C2F(recu).rstk[C2F(recu).pt - 1] = 110;
	/* *call* lprim */
	goto L100;
L86:
	op = C2F(recu).pstk[C2F(recu).pt - 1] % 256;
	kount = C2F(recu).pstk[C2F(recu).pt - 1] / 256;
	--C2F(recu).pt;
	if (op == 0) goto L82;
	Fin = op;
	Rhs = 1;
	++C2F(recu).pt;
	C2F(recu).pstk[C2F(recu).pt - 1] = kount;
	C2F(recu).rstk[C2F(recu).pt - 1] = 111;
	C2F(recu).icall = 4;
	/* *call* allops(not) */
	return 0;
L87:
	kount = C2F(recu).pstk[C2F(recu).pt - 1];
	--C2F(recu).pt;
	/* next two lines to handle a+~b and a*~b,... */
	if (C2F(recu).rstk[C2F(recu).pt - 1] == 115) goto L23;
	if (C2F(recu).rstk[C2F(recu).pt - 1] == 204) return 0;
	goto L82;
	/* in-line lprim */
L100:
	if (C2F(iop).ddt == 4) { }
	if (C2F(recu).pstk[C2F(recu).pt - 1] % 256 != 0) goto L101;
	if (C2F(com).sym != et && C2F(com).sym != ou) goto L103;
	C2F(getsym)();
	/* modif SS */
	if (C2F(eptover)(&val, &checkvalue)) return 0;
L101:
	++C2F(recu).pt;
	C2F(recu).rstk[C2F(recu).pt - 1] = 112;
	C2F(recu).icall = 1;
	/* *call* expr */
	goto L2;
L102:
	--C2F(recu).pt;
	if ((C2F(com).sym != equal && C2F(com).sym < less) || C2F(com).sym == eol) goto L86;
L103:
	op = C2F(com).sym;
	C2F(getsym)();
	if (op == equal && C2F(com).sym != equal) 
	{
		int code_message = 7;
		C2F(msgs)(&code_message, &val);
	}
	
	if (C2F(com).sym == equal || C2F(com).sym == great) 
	{
		if (op != equal) op += C2F(com).sym;
		C2F(getsym)();
		if (op == not + equal) op = less + great;
	}
	++C2F(recu).pt;
	C2F(recu).rstk[C2F(recu).pt - 1] = 113;
	C2F(recu).pstk[C2F(recu).pt - 1] = op;
	C2F(recu).icall = 1;
	/* *call* expr */
	goto L2;
L104:
	Fin = C2F(recu).pstk[C2F(recu).pt - 1];
	Rhs = 2;
	C2F(recu).rstk[C2F(recu).pt - 1] = 114;
	C2F(recu).icall = 4;
	/* *call* allops(fin) */
	return 0;
}
/*--------------------------------------------------------------------------*/ 


